File: dashboard/backend/routers/watchlist.py
"""Watchlist management API."""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from ..database import (
    upsert_watchlist_item,
    get_watchlist_items,
    delete_watchlist_item,
    get_watchlist_prices,
)

router = APIRouter()


class WatchlistItem(BaseModel):
    asset_class: str
    symbol: str
    exchange: str = ""
    notes: str = ""


class WatchlistPriceItem(BaseModel):
    id: int
    asset_class: str
    symbol: str
    exchange: str
    notes: str
    raw_data: str | None
    fetched_at: str | None


@router.get("/", response_model=list[WatchlistItem])
async def list_watchlist(asset_class: str = ""):
    """List all watchlist items, optionally filtered by asset class."""
    return await get_watchlist_items(asset_class or "")


@router.get("/prices", response_model=list[WatchlistPriceItem])
async def list_watchlist_prices(asset_class: str = ""):
    """List watchlist items with their latest cached prices."""
    return await get_watchlist_prices(asset_class or "")


@router.post("/", response_model=dict)
async def add_watchlist_item(item: WatchlistItem):
    """Add or update a watchlist item."""
    valid_classes = ("stock", "crypto", "forex", "options", "sports_card")
    if item.asset_class not in valid_classes:
        raise HTTPException(400, f"asset_class must be one of {valid_classes}")
    row_id = await upsert_watchlist_item(
        asset_class=item.asset_class,
        symbol=item.symbol,
        exchange=item.exchange,
        notes=item.notes,
    )
    return {"id": row_id, "added": True}


@router.delete("/{item_id}", response_model=dict)
async def remove_watchlist_item(item_id: int):
    """Remove a watchlist item by its ID."""
    deleted = await delete_watchlist_item(item_id)
    if not deleted:
        raise HTTPException(404, "Watchlist item not found")
    return {"deleted": True}

File: dashboard/frontend/src/api/watchlist.ts
import api from './client'

export interface WatchlistItem {
  id: number
  asset_class: 'stock' | 'crypto' | 'forex' | 'options' | 'sports_card'
  symbol: string
  exchange: string
  notes: string
  created_at?: string
}

export interface WatchlistPriceItem extends WatchlistItem {
  raw_data: { price: number; prev_close?: number; last_recorded?: number } | null
  fetched_at: string | null
}

export const watchlistApi = {
  list: () => api.get<WatchlistItem[]>('/watchlist/'),
  listWithPrices: () => api.get<WatchlistPriceItem[]>('/watchlist/prices'),
  add: (item: Omit<WatchlistItem, 'id' | 'created_at'>) =>
    api.post<{ id: number; added: boolean }>('/watchlist/', item),
  remove: (id: number) =>
    api.delete<{ deleted: boolean }>(`/watchlist/${id}`),
}
